tests/js/legendTests.js

```diff
@@ -0,0 +1,232 @@
+https://github.com/gpii/universal/LICENSE.txt
+*/
+
+(function ($, fluid) {
+
+    "use strict";
+
+    fluid.registerNamespace("gpii.tests.chartAuthoring");
+
+    fluid.defaults("gpii.tests.chartAuthoring.pieChart.legend", {
+        gradeNames: ["gpii.chartAuthoring.pieChart.legend", "autoInit"],
+        legendOptions: {
+            colors: ["#000000", "#ff0000", "#00ff00", "#0000ff", "#aabbcc", "#ccbbaa"]
+        },
+        listeners: {
+            "onLegendCreated.addMouseoverListener": {
+                listener: "{that}.addD3Listeners",
+                args: ["{that}.dom.row", "mouseover", "gpii.tests.chartAuthoring.mouseOverListener"]
+            }
+        },
+        members: {
+            mouseOverListenerCalled: false
+        }
+    });
+
+    gpii.tests.chartAuthoring.objectArray = [{
+        id: "id0",
+        value: 15,
+        label: "One"
+    }, {
+        id: "id1",
+        value: 67,
+        label: "Two"
+    }, {
+        id: "id2",
+        value: 20,
+        label: "Three"
+    }, {
+        id: "id3",
+        value: 45,
+        label: "Four"
+    }];
+
+    gpii.tests.chartAuthoring.objectArrayAdd = [{
+        id: "id0",
+        value: 15,
+        label: "One"
+    }, {
+        id: "id1",
+        value: 67,
+        label: "Two"
+    }, {
+        id: "id2",
+        value: 20,
+        label: "Three"
+    }, {
+        id: "id3",
+        value: 45,
+        label: "Four"
+    },
+    {
+        id: "id4",
+        value: 26,
+        label: "Five"
+    }];
+
+    gpii.tests.chartAuthoring.objectArrayRemove = [{
+        id: "id0",
+        value: 15,
+        label: "One"
+    }, {
+        id: "id1",
+        value: 67,
+        label: "Two"
+    }, {
+        id: "id2",
+        value: 20,
+        label: "Three"
+    }];
+
+    gpii.tests.chartAuthoring.objectArrayChangeInPlace = [{
+        id: "id0",
+        value: 36,
+        label: "I"
+    }, {
+        id: "id1",
+        value: 67,
+        label: "II"
+    }, {
+        id: "id2",
+        value: 26,
+        label: "Three"
+    }];
+
+    gpii.tests.chartAuthoring.mouseOverListener = function (data, i, that) {
+        that.mouseOverListenerCalled = true;
+    };
+
+    // convenience function for easing testing of colors (jquery returns only RGB)
+    // Based off http://stackoverflow.com/questions/4262417/jquery-hex-to-rgb-calculation-different-between-browsers
+    gpii.tests.chartAuthoring.hexToRGB = function (hexStr){
+        // note: hexStr should be #rrggbb
+        var hex = parseInt(hexStr.substring(1), 16);
+        var r = (hex & 0xff0000) >> 16;
+        var g = (hex & 0x00ff00) >> 8;
+        var b = hex & 0x0000ff;
+        return "rgb(" + r + ", " + g + ", " + b + ")";
+    };
+
+    gpii.tests.chartAuthoring.testLegendSyncWithModelDataSet = function(that) {
+        var dataSet = that.model.dataSet;
+        //
+        if(that.options.legendOptions.sort) {
+            dataSet.sort(function(a,b) {
```

Comment:
It's actually that we need to sort the model's dataSet in test in ascending order to compare it against the rendered components, if the sort option is true. At the moment, legend.js leaves the original dataset's ordering alone, but sorts the D3-based elements when sort: true.

It is possible this is overly complicated & we should simply sort the dataSet whenever it's updated if the sort option is true in legend.js. I think my logic was that you may want to preserve original dataset order for other reasons while sorting the visual display.